usvm-ts/src/main/kotlin/org/usvm/machine/expr/TsLocalResolver.kt

```diff
@@ -0,0 +1,199 @@
+package org.usvm.machine.expr
+
+import io.ksmt.utils.cast
+import mu.KotlinLogging
+import org.jacodb.ets.model.EtsLexicalEnvType
+import org.jacodb.ets.model.EtsLocal
+import org.jacodb.ets.model.EtsParameterRef
+import org.jacodb.ets.model.EtsThis
+import org.jacodb.ets.model.EtsValue
+import org.jacodb.ets.utils.DEFAULT_ARK_METHOD_NAME
+import org.usvm.UExpr
+import org.usvm.api.allocateConcreteRef
+import org.usvm.dataflow.ts.infer.tryGetKnownType
+import org.usvm.machine.TsContext
+import org.usvm.machine.interpreter.TsStepScope
+import org.usvm.machine.interpreter.getGlobals
+import org.usvm.machine.interpreter.initializeGlobals
+import org.usvm.machine.interpreter.isGlobalsInitialized
+import org.usvm.machine.state.TsMethodResult
+import org.usvm.memory.ULValue
+import org.usvm.util.SymbolResolutionResult
+import org.usvm.util.mkFieldLValue
+import org.usvm.util.mkRegisterStackLValue
+import org.usvm.util.resolveImportInfo
+
+private val logger = KotlinLogging.logger {}
+
+fun TsContext.resolveLocal(
+    scope: TsStepScope,
+    local: EtsValue,
+): UExpr<*>? {
+    check(local is EtsLocal || local is EtsThis || local is EtsParameterRef) {
+        "Expected EtsLocal, EtsThis, or EtsParameterRef, but got ${local::class.java}: $local"
+    }
+
+    // Handle closures
+    if (local is EtsLocal && local.name.startsWith("%closures")) {
+        // TODO: add comments
+        val existingClosures = scope.calcOnState { closureObject[local.name] }
+        if (existingClosures != null) {
+            return existingClosures
+        }
+        val type = local.type
+        check(type is EtsLexicalEnvType)
+        val obj = allocateConcreteRef()
+        // TODO: consider 'types.allocate'
+        for (captured in type.closures) {
+            val resolvedCaptured = resolveLocal(scope, captured) ?: return null
+            val lValue = mkFieldLValue(resolvedCaptured.sort, obj, captured.name)
+            scope.doWithState {
+                memory.write(lValue, resolvedCaptured.cast(), guard = trueExpr)
+            }
+        }
+        scope.doWithState {
+            setClosureObject(local.name, obj)
+        }
+        return obj
+    }
+
+    val lValue = resolveLocalToLValue(scope, local) ?: return null
+    return scope.calcOnState { memory.read(lValue) }
+}
+
+fun TsContext.resolveLocalToLValue(
+    scope: TsStepScope,
+    local: EtsValue,
+): ULValue<*, *>? {
+    val currentMethod = scope.calcOnState { lastEnteredMethod }
+
+    if (currentMethod.name == DEFAULT_ARK_METHOD_NAME) {
+        // TODO
```

Comment:
Add at least a comment, what to do here, and what this section means

Reply:
WIP

Reply:
Done. It was required to adopt the reading of "locals" in `%dflt::%dflt`, similar to how we ad-hoc the assignments to those "locals": if the "local" variable _looks like a global_, then we should treat it as global, not local. In the perfect world, `method.body.locals` would contain the correct list of "real locals", and this ad-hoc would not be needed, but currently the IR is broken (see issue https://gitcode.com/openharmony-sig/arkanalyzer/issues/768).

For now, the following heuristic works for determining whether the variable "looks like a global". The only "real local" variable I have seen are: "temp locals" (`%0` from AA, and `_tmp0` from our 3AC builder) and the local named "this". All other variables in `%dflt` are considered global, both when reading AND writing them.